### PR TITLE
Fix string quoting/escaping in message and properties

### LINKED ISSUE #117: JSON quotes are escaped as `\"`
A json string was logged as below in v2.6.2:
```
{"prop":"value"}
```

But it's logged (looks like escaped) as below in v3.0.3:
```
{\"prop\":\"value\"}
```

This is in the `Custom Properties`. **To clarify**, I'm logging json request/response body as string. 

This looks like a bug to me. Is this intended?

### LINKED ISSUE #137: Allow not escaping string (the :l flag on {Message}) (or allow setting the output template)
(Related to #117, but this issue has another focus.)

As you are probably aware, Serilog allows [setting custom output templates](https://github.com/serilog/serilog/wiki/Formatting-Output#formatting-plain-text). The `{Message}` token supports a flag `:l` that turns off quoting of strings.

I am able to configure the output template for all my sinks except Serilog.Sinks.ApplicationInsights. It seems that Serilog.Sinks.ApplicationInsights does not use the `:l` flag, because when my message template contains a JSON string, it is escaped.

My message template is this:

```
Log.Debug("Operation {OpName}\n{InteractionData}", opName, json)
```

My wish is to have a rendered message that looks like this:

```
Operation User.byId
{
  "Input": {
    "UserId": 10036
  },
  "Output": null
}
```

Here is how the message looks in AI:

image

If I view just the telemetry property corresponding to the JSON in AI, it looks even worse:

image

This looks similar to how the console sink looks when just using `{Message}` (no `:l` flag):

image

Whereas if I specify `{Message:l}` in the output template for the console sink, it looks just like I want it to:

image

My guess is that if Serilog.Sinks.ApplicationInsights were to allow setting a custom output template, and I then use `{Message:l}`, then the message and/or the property in AI would look properly formatted, as shown in the last console image above.

### PR DESCRIPTION
**Pull request by cmeeren**
Fixes #137
Fixes #117

Need help, currently not working. The proposed solution from https://github.com/serilog-contrib/serilog-sinks-applicationinsights/issues/137#issuecomment-1115434026 wasn't right, it seems. Or I misunderstood it. When I run the new tests, then the `includeRenderedMessage` in that method is `false`, so the modified code isn't called.

For consideration:
* The most important part for me is the property data itself, though it would be nice to have the option of not quoting message strings, too.
* Should this be configurable? If so, I need guidance.

### CODE REVIEW

**Comment by nblumhardt** [comment]
Thanks for taking a look, @cmeeren !

RE not quoting message strings, this is another change I think we should make; https://github.com/serilog-contrib/serilog-sinks-applicationinsights/issues/117 is tracking it, I believe. Nice to roll in, if there's an obvious place to make the change.

I'm taking a look at the tests now 👍

**Comment by nblumhardt** [comment]
Think I've figured it out.. We need a new unit test:

```csharp
// EventTelemetryConverterTest.cs
using Serilog.Sinks.ApplicationInsights.TelemetryConverters;
using Xunit;

namespace Serilog.Sinks.ApplicationInsights.Tests
{
    public class EventTelemetryConverterTest : ApplicationInsightsTest
    {
        public EventTelemetryConverterTest() : base(new EventTelemetryConverter())
        {
        }

        [Fact]
        public void MessagesAreFormattedWithoutQuotedStrings()
        {
            Logger.Information("Hello, {Name}!", "world");
            Assert.Equal("Hello, world!", LastSubmittedEventTelemetry.Properties["RenderedMessage"]);
        }
    }
}
```

This fails for me now with:

```
Assert.Equal() Failure
                 ↓ (pos 7)
Expected: Hello, world!
Actual:   Hello, "world"!
                 ↑ (pos 7)
   at Serilog.Sinks.ApplicationInsights.Tests.EventTelemetryConverterTest.MessagesAreFormattedWithoutQuotedStrings() in ...
```

I added `LastSubmittedEventTelemetry` to the test base class; it's:

```csharp
        protected EventTelemetry LastSubmittedEventTelemetry => 
            _channel.SubmittedTelemetry
                .OfType<EventTelemetry>()
                .LastOrDefault();
```

HTH!

**Comment by cmeeren** [comment]
Ok, tests are now improved and I have separate tests for trace and event telemetry. But the suggested implementation from https://github.com/serilog-contrib/serilog-sinks-applicationinsights/issues/137#issuecomment-1115434026 does not change the result of any of the tests.

I'm not sure where the message is being created; do you know where to make the actual change? And I suppose I should revert the change you suggested?

**Comment by cmeeren** [comment]
Update: My IDE/test runner seems to have some issues. To correct what I said earlier, the suggested implementation makes two of the tests pass:
* `EventTelemetryConverterTest.MessagesAreFormattedWithoutQuotedStrings`
* `EventTelemetryConverterTest.MessageQuotesAreNotEscaped`

However, the last EventTelemetry test for message property values, as well as all three tests for trace telemetry, are failing.

**Comment by cmeeren** [comment]
Managed to find the place to edit trace telemetry.

Will now try to figure out the message properties. Any help welcome.

**Comment by cmeeren** [comment]
Ok, so the formatting of the properties come from Serilog's `JsonValueFormatter` through this repo's `ApplicationInsightsJsonValueFormatter`. This is documented [in this repo's readme](https://github.com/serilog-contrib/serilog-sinks-applicationinsights/tree/44839c95ebdda74e949117f90219f7535dc53d27#how-custom-properties-are-logged).

From what I can gather, issue #59 was where this was decided, and [this blog post of yours from 2016](https://nblumhardt.com/2016/07/serilog-2-0-json-improvements/) has more details.

I think perhaps it makes most sense for us to keep the current behavior (JSON formatting, i.e. escaping quotes) for property values. Personally, as long as the message itself is rendered without quotes and therefore more readable than now, I can live with it.

Just to mention it, an alternative is to only format the properties as JSON when destructuring (using `@`), and write the value directly (not format it as JSON) if not destructured. This would require documenting the migration path for people who would like to keep the current behavior, which is either to use `@` for all properties they want formatted as a complete, valid JSON document, or, assuming we add another `IValueFormatter` that is used by default in v4, to revert to `ApplicationInsightsJsonValueFormatter`.

**Comment by cmeeren** [comment]
Hmm, I may have changed my opinion on the current behavior.

I currently have no good solution for destructuring in F# (including Destructurama.FSharp, see https://github.com/destructurama/fsharp/issues/28). Therefore I manually serialize objects when using them in payload properties. 

Here is an example of a serialized payload string (simplified):

```
{
  Input: {
    ProjectId: 10002
  },
  Output: {
    Project: 10002
    }
  }
}
```

Here is how it's currently rendered in the message:

```
Interaction "ProjectSettings.byId"
"{
  \"Input\": {
    \"ProjectId\": 10002
  },
  \"Output\": {
    \"Project\": 10002
    }
  }
}"
```

Here is the raw property value that is sent to AI:

```
{\r\n  \"Data\": \"This is a test\",\r\n  \"Sub\": {\r\n    \"Bar\": [\r\n      2\r\n    ],\r\n    \"Foo\": {\r\n      \"Case\": \"Ok\",\r\n      \"ResultValue\": 1\r\n    }\r\n  }\r\n}
```

In other words, it has taken my existing serialized JSON, and formatted it as a JSON string token, escaping newlines and quotes.

I would like a way to send my JSON as-is, without it being transformed.

(Ideally I'd like a way to destructure F# types properly in Serilog, but that is further down the road.)

**Comment by cmeeren** [comment]
My previous comment is no longer be relevant if using the `IDestructuringPolicy` posted in  https://github.com/serilog/serilog/issues/1675. If destructuring instead of logging a manually serialized JSON string, then instead of e.g.

```
{\r\n  \"Data\": \"This is a test\",\r\n  \"Sub\": {\r\n    \"Bar\": [\r\n      2\r\n    ],\r\n    \"Foo\": {\r\n      \"Ok\": 1\r\n    }\r\n  }\r\n}
```

I get

```
{"Data":"This is a test","Sub":{"Bar":[2],"Foo":{"Ok":1}}}
```

So in terms of properties, my use-case is solved when destructuring using the mentioned `IDestructuringPolicy`.

**Comment by cmeeren** [comment]
Actually, I'm not even sure about the change already in this PR.

Using the destructuring linked to above, if I call

```f#
let data = {| Data = "This is a test"; Sub = {| Foo = Ok 1; Bar = [Some 2] |} |}
Log.Information("Data: {@Data}", data)
```

Then with the current published version of Serilog.Sinks.ApplicationInsights, I get this in the raw AI payload:

```json
"message": "Data: { Data: \"This is a test\", Sub: { Bar: [2], Foo: { Ok: 1 } } }"
```

That looks like exactly what I want: No escaped strings (singly escaped above because it's part of the AI payload).

With the changes in this PR, I get:

```json
"message": "Data: {\"Data\":\"This is a test\",\"Sub\":{\"Bar\":[2],\"Foo\":{\"Ok\":1}}}",
```

So for destructuring, it looks like this PR as it currently stands is a step in the wrong direction.

I'm a bit confused right now and may be missing something. Please advise.

**Comment by cmeeren** [comment]
I have now pushed two more tests to verify the format I'd like in messages when destructuring objects. They pass with the old implementation, but not the new one.

I welcome ideas on how to make all current tests pass (or, alternatively, an explanation of why that is futile or a should not be a goal).

**Comment by cmeeren** [comment]
Oh wow, so that was easy.

**Comment by cmeeren** [comment]
@nblumhardt, do you know why, with the current implementation, this:

```f#
type A(b: string, c: int) =
  member val B = b
  member val C = c

type B(a: A) =
  member val A = a

let data = B(A("foobar", 123))
Log.Information("Data: {@Data}", data)
```

Renders like:

```
Data: B { A: A { B: "foobar", C: 123 } }
```

(which is what I want), but Serilog.Sinks.Console (using the `:l` switch and not the `:j` switch, just like here) renders as

```
Data: B {A=A {B="foobar", C=123}}
```

In other words, it seems like Serilog.Sinks.ApplicationInsights, even without the `:j` (JSON formatting) flag, renders it like "less verbose JSON" (property names not quoted), whereas the console sink renders it using a different style?

**Comment by nblumhardt** [comment]
@cmeeren thanks for pushing this forwards! I think the PR as it currently stands is an improvement and we should probably call it a checkpoint and merge as-is 👍 

Regarding the various formats you see with `@` and `:l` etc, the only really consistent one is the JSON version, and most times I think `:lj` is the way to go, now.

For property values, the intention of #59 seems to have been to handle arrays and structures more gracefully, but in the process, accidentally changed behavior for regular strings as well. This should be able to be fixed/worked around fairly easily. I'll make sure #117 stays open, and take a look at that separately.

Thanks again for your help investigating and improving this - much needed by this project :-) 🙇
